### PR TITLE
Rec system live version 2

### PR DESCRIPTION
**Pull request by Iamogeee**
This pull request implements a recommendation system using a linear regression model built from scratch, without relying on the previous external library, TensorFlow.js. The changes made in the code focus on leveraging JavaScript and custom functions to achieve a content-based filtering recommendation system.

**Changes Implemented**
Feature Engineering: 
- Created feature vectors for each business and user by combining their attributes.

Model Training: 
- Implemented a linear regression model from scratch using native JavaScript.
- Trained the model using the combined feature vectors and interaction ratings as target labels.
- Trained the model using gradient descent to minimize the error between predicted and actual ratings.

Prediction and Recommendation Generation: 
- Predicted user preferences for businesses using the trained model.
- Ranked businesses based on predicted ratings.
- Generated top recommendations and fetched detailed business information for the recommended businesses.

Future implementations:
Continue advanced feature engineering to increase the complexity and improve the recommendation system's accuracy.

Demo:
<div>
    <a href="https://www.loom.com/share/b29a0b4bf25a480485ebf71fe5d9368b">
      <p>Loom Message - 15 July 2024 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/b29a0b4bf25a480485ebf71fe5d9368b">
      
    </a>
  </div>

### CODE REVIEW

File: backend/recommendationSystem.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      normalizeRating(business.averageRating || 0),
+      ...oneHotEncode([business.category], allCategories),
+    ],
+  }));
+
+  // Train the model
+  const model = await trainLinearRegression([user], businessData);
+
+  // Save the model to a file
+  fs.writeFileSync("model.json", JSON.stringify(model, null, 2));
```
Comment (sinap-fb) [inline backend/recommendationSystem.js:182]: should this model file be specific to the user? So that when different users hit this code, they don't overwrite each other's `model.json`.  One way of achieving this could be to make name the file something like `model_{user_id}.json`. (And doing the same for wherever we read `model.json`)

**Review by sinap-fb — COMMENTED** [review]
LGTM, but i can't really speak to the ML stuff

File: backend/recommendationSystem.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+      y.push(rating);
+    });
+  });
+
+  const numFeatures = X[0].length;
+  let weights = Array.from({ length: numFeatures }, () => Math.random() * 0.01);
+  let bias = Math.random() * 0.01;
+
+  for (let epoch = 0; epoch < epochs; epoch++) {
```
Comment (mylesdomingo) [inline backend/recommendationSystem.js:121]: Can you link any math sources to show how you calculated the predicition model?

**Comment by mylesdomingo** [comment]
The UI is a bit unclear, in the recommendations tab are all these listings recommended by the algorithm? 
How many businesses are in our data set? It doesn't seem like we have that many to provided more than 1 or 2 recommnedations.

File: backend/recommendationSystem.js (hunk 2 above)
Comment (Iamogeee) [reply backend/recommendationSystem.js:121]: Sure! I'll do this in the next PR, cause I made some changes

File: backend/recommendationSystem.js (hunk 1 above)
Comment (Iamogeee) [reply backend/recommendationSystem.js:182]: Yes, the model is specific to the user. However, I would like some more info about this approach if you don't mind

**Comment by Iamogeee** [comment]
> The UI is a bit unclear, in the recommendations tab are all these listings recommended by the algorithm? How many businesses are in our data set? It doesn't seem like we have that many to provided more than 1 or 2 recommendations.

@mylesdomingo I didn't want the recommendations page to be scanty at first glance for a new user (this was something Jose also agreed on). So at first, based on the user's favorite categories and preferred rating, the algorithm produces recs from all categories. Then as the user continually interacts with the website, the model retrains and gives more accurate recs.
As for the data set, I do still plan on seeding in more data and categories.

File: backend/recommendationSystem.js (hunk 1 above)
Comment (sinap-fb) [reply backend/recommendationSystem.js:182]: Gotcha. So for instance this line could be changed to something like
```js
  fs.writeFileSync(`model_${user_id}.json`, JSON.stringify(model, null, 2));
```
This means that when you retrain the model (e.g. switch to another account), the file name will be e.g. `model_2.json`. Then when a user makes a request that reads the model JSON file, we'd use the same sort of filename: `model_${user_id}.json`

This way, if you are logged in to user **2**, the server reads/writes from the file `model_2.json`.
When you switch to user **3**, the server reads/writes from the file `model_3.json`
